### PR TITLE
CalcSpatialInertia handles Convex correctly

### LINKED ISSUE #21666: CalcSpatialInertia fails on this Hello Robot obj
### What happened?

Following #20444, I've put a simple failure reproduction here: https://github.com/RobotLocomotion/drake/commit/e470b302db69daf252877bee166ef4160a3e31be

This attempts to compute the spatial inertia from the hello robot model in mujoco menagerie but fails with
```
Expected: CalcSpatialInertia(unit_scale_obj, kDensity) doesn't throw an exception.
  Actual: it throws std::runtime_error with description "Spatial inertia fails SpatialInertia::IsPhysicallyValid().
 mass = 2.1647959683444456e-05
 Center of mass = [-0.24465954725602354  -3.041901842037344e-06  0.02583678929053868]
 Inertia about point P, I_BP =
[1.90172e-08  -1.2568e-11  1.35882e-07]
[-1.2568e-11  1.30846e-06  1.45236e-12]
[1.35882e-07  1.45236e-12   1.2914e-06]
 Inertia about center of mass, I_BBcm =
[ 4.56633e-09   3.54305e-12  -9.59847e-10]
[ 3.54305e-12  -1.79939e-09  -2.49019e-13]
[-9.59847e-10  -2.49019e-13  -4.40846e-09]
 Principal moments of inertia about Bcm (center of mass) =
[-4.509963383944518e-09  -1.799395202505969e-09  4.667840740414944e-09]
```

Additionally the `CalcSpatialInertia` doc strings do not actually declare that they could throw an exception for this reason.

### Version

_No response_

### What operating system are you using?

Ubuntu 22.04

### What installation option are you using?

compiled from source code using Bazel

### Relevant log output

_No response_

### LINKED ISSUE #21681: CalcSpatialInerta(Shape, double) doesn't respect Convex
### What happened?

Shape specification promises that everywhere `Convex` is used the convex hull of the mesh is used and not the mesh itself. In `geometry_spatial_inertia.{h|cc}`, `CalcSpatialInertia` does not support this promise. It treats the `Convex` the same as the `Mesh` and attempts to compute spatial inertia for the mesh and not its convex hull. It needs to be brought in line with the documented promise.

### Version

_No response_

### What operating system are you using?

_No response_

### What installation option are you using?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by SeanCurtis-TRI**
`CalcSpatialInertia()` was trying to operate directly on the mesh file, counter to the documented contract that a `Convex` shape is *always* its convex hull.

Updates mujoco parser to use the convex hull's spatial inertia when attempts to compute *a* physically valid spatial inertia for a mesh fails. Note: the warning stays (but is updated). The mujoco parser likewise posts a warning when the mesh is replaced by its convex hull.

Resolves #21681

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)

### CODE REVIEW

**Review by SeanCurtis-TRI — COMMENTED** [review]
+(release notes: fix) 
+@RussTedrake for feature review, please.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-:-O1Igt5-88tRMpX47GcK:b-sfk2wf)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: 1 unresolved discussion, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 435 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1It0hn1KrQ8EpHtkCH:-O1It0hn1KrQ8EpHtkCI:b-27auy5) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L435)):*
> ```cpp
>         M_GG_G_ = mesh_inertia_->at(name_);
>       } else {
>         try {  // TODO(21666), remove the try-catch.
> ```

BTW Leaving the TODO intact for an issue number this PR claims to fix is not a great look.

I imagine you will want something like this comment, instead:

https://github.com/RobotLocomotion/drake/blob/78a2f1c2192dc941136652c666d5a66ea96da11f/multibody/parsing/detail_common.cc#L304-L311

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 435 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1It0hn1KrQ8EpHtkCH:-O1IwU8yEVbYeHHMq8hx:b-p9n2zw) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L435)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Leaving the TODO intact for an issue number this PR claims to fix is not a great look.

I imagine you will want something like this comment, instead:

https://github.com/RobotLocomotion/drake/blob/78a2f1c2192dc941136652c666d5a66ea96da11f/multibody/parsing/detail_common.cc#L304-L311
</blockquote></details>

Oops...I need to back out my claim. That issue has *two* components:

1) We're doing the right thing with mesh spatial inertia (accomplished by this PR).
2) We're doing it the right way (not accomplished).

I'm going to retract the statement.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 435 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1It0hn1KrQ8EpHtkCH:-O1IxDf27XnzFcz_oA-x:b-pv9y7j) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L435)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Oops...I need to back out my claim. That issue has *two* components:

1) We're doing the right thing with mesh spatial inertia (accomplished by this PR).
2) We're doing it the right way (not accomplished).

I'm going to retract the statement.
</blockquote></details>

Fine by me, thanks.

**Review by RussTedrake — COMMENTED** [review]
Thanks Sean. I can probably get to this in a day or two, but have to focus on another task first...

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-:-O1RLsFK-pBw0ry_BRej:b-6m2syy)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers

**Review by RussTedrake — COMMENTED** [review]
:lgtm:

+@jwnimmer-tri for platform review, please.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-:-O1g0rWB6tKLjcpHr1zU:b1f2m6h)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1g0ujODVnfxNzR1nf_):*
nit: directly.

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: in the big picture, but several details still need attention.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-:-O1gwVup6sWaUNyCqD8j:b-7pwlbq)* status: 4 unresolved discussions

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1guLh43TNtw8XtAlih:-O1guLh51jng-KJBQceP:bp39p6t) (no related file):*
In `geometry_spatial_inertia.h`, it currently says this:

>  Note: Spatial inertia calculations for the geometry::Convex type do not currently require that the underlying mesh actually be convex. Although certain collision calculations involving geometry::Convex may use the mesh's convex hull, this function uses the *actual* mesh.

I think we need to update (or remove?) that disclaimer, now.

___
___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1gvSen2GQBiPrddFlE:b-cy2fze) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
> ```cpp
>       const SpatialInertia<double> M_GG_G_one = calculator.Calc(*geom.shape);
>       if (calculator.used_convex_hull_fallback()) {
>         Warning(
> ```

In our philosophy of parsing, if Drake is _correctly_ implementing the upstream specification, then we should not print a warning.  In that sense, I think (at least for some input files) this warning is a false positive?  In some cases, the _spec_ is that we are supposed to fill up the geometry to its convex hull during inertia calc, I think?  At least, that's what I recall from the issue threads.

Having said that... I could see a valid case for eating some false positives here, in order to avoid false negatives where we would silently mis-parse a file that upstream is _not_ using the hull for.

In any case, this question probably deserves some very careful comments in this file explaining the situation for future readers.  The key point would be any gap between which geoms _we_ decide need the hull fallback, vs what geoms MuJoCo decides need the hull fallback.  This issue has had a lot of discussion, and we need to reflect a summary of that back into the code here, for next time around.
___
*[`multibody/tree/test/geometry_spatial_inertia_test.cc` line 203 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1guyU65S7oQfCrC2GE:-O1guyU65S7oQfCrC2GF:b-td3py2) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/tree/test/geometry_spatial_inertia_test.cc#L203)):*
Somehow I don't quite follow this.

The call to action atop this test stanza seems right to me:

> ... .obj and .vtk don't throw, everything else does ...

We still want Convex to throw, don't we?  (And therefore to test that it throws.)

Is the problem here that it has a different error message for Convex vs Mesh?  That seems solveable.

Or if the thought is that some other unit test in our tree sufficiently covers this logic, we could cite that here in a comment to explain why we can skip this test case.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1qr7b3AkqbU6tNL8rB:b-7fgbnr) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
>Mujoco prints a warning when it swaps convex hull for mesh. I'd argue that printing a warning in this case is "correctly implementing the upstream specification".

Works for me.  I hadn't remembered that from the issue stream.

> ... it is purely by virtue of reading the code.

That's just now MuJoCo rolls.  I think using its code as its specification is the best we can do.  (I don't think limiting ourselves to "only what appears in the documentation" is a net win.)

> There exists a class of meshes for which it is possible that Mujoco and Drake disagree on whether the computed spatial inertia is physically valid

This part is actually is my main worry.  I would like to see somewhere in writing (maybe as a final thought summary post the Drake issue) saying that because MuJoCo doesn't properly specify the preconditions for convex fallback, it's impossible for us to precisely implement the same specification, and therefore our parser and their parser will always risk a disagreement.  At least both of them warn, so if the users sees a different warning landscape across the two parsers, at least they have a clue about why their physics differs in each simulator.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1quZClE9rpgko4yMGR:b-rlfnm9) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

>Mujoco prints a warning when it swaps convex hull for mesh. I'd argue that printing a warning in this case is "correctly implementing the upstream specification".

Works for me.  I hadn't remembered that from the issue stream.

> ... it is purely by virtue of reading the code.

That's just now MuJoCo rolls.  I think using its code as its specification is the best we can do.  (I don't think limiting ourselves to "only what appears in the documentation" is a net win.)

> There exists a class of meshes for which it is possible that Mujoco and Drake disagree on whether the computed spatial inertia is physically valid

This part is actually is my main worry.  I would like to see somewhere in writing (maybe as a final thought summary post the Drake issue) saying that because MuJoCo doesn't properly specify the preconditions for convex fallback, it's impossible for us to precisely implement the same specification, and therefore our parser and their parser will always risk a disagreement.  At least both of them warn, so if the users sees a different warning landscape across the two parsers, at least they have a clue about why their physics differs in each simulator.
</blockquote></details>

I may have inspired unintended concerns in you. This isn't really about specification. It's about floating point calculation.

Volume is computed with respect to an arbitrary point. Drake uses the origin of the mesh's frame (because it simplifies the algebra and assumes most meshes are defined near the origin). Mujoco computes and uses a "face centroid". So, the disparity I was alluding to -- purely out of legal pedantry -- is attributable to floating point error in the different formulations. I don't think it would appreciably impact anyone. If someone's mesh has a volume so close to zero that the difference in formulation causes the volume to become negative in one implementation vs the other, they surely have other problems already.

(This is based on my belief that a test for a physically valid inertia is correlated with a negative volume -- Drake's test vs Mujoco's test.)

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1r7_bCAcVw9FPM0np9:b-payl6g) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I may have inspired unintended concerns in you. This isn't really about specification. It's about floating point calculation.

Volume is computed with respect to an arbitrary point. Drake uses the origin of the mesh's frame (because it simplifies the algebra and assumes most meshes are defined near the origin). Mujoco computes and uses a "face centroid". So, the disparity I was alluding to -- purely out of legal pedantry -- is attributable to floating point error in the different formulations. I don't think it would appreciably impact anyone. If someone's mesh has a volume so close to zero that the difference in formulation causes the volume to become negative in one implementation vs the other, they surely have other problems already.

(This is based on my belief that a test for a physically valid inertia is correlated with a negative volume -- Drake's test vs Mujoco's test.)
</blockquote></details>

I wasn't worried about FP.  Is the MuJoCo "is valid" predicate *really* exactly the same as Paul's predicate in Drake?  MuJoCo also checks the triangle inequality?  And when using the same Body frame to shift the inertia into?

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1rBW0E-YOlesUyM-YM:bsk9gpm) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I wasn't worried about FP.  Is the MuJoCo "is valid" predicate *really* exactly the same as Paul's predicate in Drake?  MuJoCo also checks the triangle inequality?  And when using the same Body frame to shift the inertia into?
</blockquote></details>

I know you weren't worried about FP. I was making a veiled allusion to FP issues, but I felt you were springboarding off of that to some non-issue. There may be larger issues of "do Drake and Mujoco do equivalent things". But for the scope of the changes in this PR, I'm quite content.

Mujoco's predicate is purely based on the sign of the computed volume. When we compute the `SpatialInertia` the total mass we report is the product of the requested density and the computed *signed* volume. So, if we compute a negative volume, we'll get a negative mass which is the *first* failure condition for `SpatialInertia::IsPhysicallyValid()`. So, I'm content with the equivalency between Drake's test and Mujoco's. (Certainly, we have additional failure modes for an invalid spatial inertia defined in `IsPhysicallyValid()`, but I don't believe those would be tickled by our inertia-from-mesh computations.)

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1rFviw8d5_VEwkfPfH:b-j6phkl) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I know you weren't worried about FP. I was making a veiled allusion to FP issues, but I felt you were springboarding off of that to some non-issue. There may be larger issues of "do Drake and Mujoco do equivalent things". But for the scope of the changes in this PR, I'm quite content.

Mujoco's predicate is purely based on the sign of the computed volume. When we compute the `SpatialInertia` the total mass we report is the product of the requested density and the computed *signed* volume. So, if we compute a negative volume, we'll get a negative mass which is the *first* failure condition for `SpatialInertia::IsPhysicallyValid()`. So, I'm content with the equivalency between Drake's test and Mujoco's. (Certainly, we have additional failure modes for an invalid spatial inertia defined in `IsPhysicallyValid()`, but I don't believe those would be tickled by our inertia-from-mesh computations.)
</blockquote></details>

Ah.  I hadn't realized that the only prong of `IsPhysicallValid` that mesh inertia computations could fail (and thus throw for) is the sign.  Maybe that belongs in a comment somewhere?  I don't think its obvious.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1rGU1JACQr8KBPBiOV:bpnniod) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ah.  I hadn't realized that the only prong of `IsPhysicallValid` that mesh inertia computations could fail (and thus throw for) is the sign.  Maybe that belongs in a comment somewhere?  I don't think its obvious.
</blockquote></details>

I'll confirm with @mitiguy that that is the case and take action accordingly.

**Review by RussTedrake — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: 1 unresolved discussion

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: 1 unresolved discussion

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1xcVnT77HW8me6CLYQ:b-xzi36d) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'll confirm with @mitiguy that that is the case and take action accordingly.
</blockquote></details>

In consultation with @mitiguy we've confirmed that `SpatialInertia::IsPhysicallyValid()` is sensitive to more failure modes than mujoco uses.

That said, while it's *possible* to create a mesh that has positive mass/volume but negative inertia, it's *tricky*. Generally, it is unlikely that Drake and Mujoco would differ on mesh vs convex for most (if not all) real world meshes.

How would you like to handle the theoretical distinction?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: 1 unresolved discussion

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1xd3ycADwUz93HKxly:bn0ojt0) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

In consultation with @mitiguy we've confirmed that `SpatialInertia::IsPhysicallyValid()` is sensitive to more failure modes than mujoco uses.

That said, while it's *possible* to create a mesh that has positive mass/volume but negative inertia, it's *tricky*. Generally, it is unlikely that Drake and Mujoco would differ on mesh vs convex for most (if not all) real world meshes.

How would you like to handle the theoretical distinction?
</blockquote></details>

Basically copy-paste those sentences into a comment inside our parser.

In shorthand:
- The recipe that Drake's MJCF parser uses to toggle on convex hull for inertia differs from the MuJoCo recipe.
- In practice, it's very unlikely that they ever disagree, so this is fine.

It doesn't need to be a TODO or anything -- just brain dump for next person who circles by here and might wonder about the mismatch.

---

For bonus points, Paul could submit a PR to MuJoCo changing their is-valid-inertia recipe in their parser to use our physically valid recipe, since I imagine ours is better.  (Then we could circle back and remove the new comment.)

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1xgbo5AGqWV9fzIgm8:b-bw8e14) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Basically copy-paste those sentences into a comment inside our parser.

In shorthand:
- The recipe that Drake's MJCF parser uses to toggle on convex hull for inertia differs from the MuJoCo recipe.
- In practice, it's very unlikely that they ever disagree, so this is fine.

It doesn't need to be a TODO or anything -- just brain dump for next person who circles by here and might wonder about the mismatch.

---

For bonus points, Paul could submit a PR to MuJoCo changing their is-valid-inertia recipe in their parser to use our physically valid recipe, since I imagine ours is better.  (Then we could circle back and remove the new comment.)
</blockquote></details>

Updated.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1xh9bz4rrKvoJrVMST:bxztp5u) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Updated.
</blockquote></details>

Beautiful.

**Review by mitiguy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21693)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform)

___
*[`multibody/parsing/detail_mujoco_parser.cc` line 797 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21693#-O1gvSen2GQBiPrddFlD:-O1xy_f60oJnX4Qh8JuT:b-ocjoh4) ([raw file](https://github.com/RobotLocomotion/drake/blob/19ce7f907b8a0a24dd848b00e267db6417c08b1f/multibody/parsing/detail_mujoco_parser.cc#L797)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Beautiful.
</blockquote></details>

I discussed the PR to MujoCO with Sherm.  Rather than trying to digest the MujoCo build & review system, it may make sense to write something up that uses the simple two tetrahedra example discussed with Sean that shows the missing validity check and points to the Drake code that does a more exhaustive validity check.   

FYI: Drake's spatial inertia validity checks are done without knowledge of the mesh.  It may be useful to do a secondary validity check (not in the Spatial Inertia class).   With the extra knowledge of the physical mesh, we can do better checking, e.g., the center of mass location should not be outside the object's bounding box and moments of inertia should not be larger than theoretically possible. There may be cases when the spatial inertia validity passes, even though the center of mass location is nonsensical or the moments of inertia are larger than possible.  Flagging nonsensical mass/inertia properties can save roboticists much time, particularly when there are units conversion issues in inertia.
